build: Add OpenBSD support.

brad0 opened the pull request.

brad0 committed:
build: Add OpenBSD support.

bradleysepos commented:
Thanks for this. Initial read looks good. I'll give it a whirl in a VM when I have some time, hopefully soon.

I suspect the dependencies are essentially the same as NetBSD? https://github.com/HandBrake/HandBrake-docs/blob/master/source/docs/en/latest/developer/install-dependencies-netbsd.markdown

brad0 commented:
> Thanks for this. Initial read looks good. I'll give it a whirl in a VM when I have some time, hopefully soon.
> 
> I suspect the dependencies are essentially the same as NetBSD? https://github.com/HandBrake/HandBrake-docs/blob/master/source/docs/en/latest/developer/install-dependencies-netbsd.markdown

Some of the names are slightly different, but that's about it.

brad0 commented:
Let me know if you need any assistance.

bradleysepos commented:
I'm not certain the best way to document the dependencies. We generally provide a human-readable list accompanied by commands suitable for direct execution in one's terminal, cf. https://handbrake.fr/docs/en/latest/developer/install-dependencies-netbsd.html.

OpenBSD's `pkg_add` generally works in interactive mode, and with many versions of packages available, I'm not aware of the preferred incantation for installing a suitable set of packages with a one liner. For example, let's install `autoconf`:

```
$ pkg_add autoconf
#...
Ambiguous: choose package for autoconf
        0: <None>
        1: autoconf-2.13p4
        2: autoconf-2.52p6
#...
        16: autoconf-2.69p3
        17: autoconf-2.71
Your choice: 
```

There are seventeen available packages and nothing here indicates `autoconf-2.69p3` is already installed. Presumably, the user should know their system well enough to make a competent decision. Specifying the full package name with version does install without further interruption, but can lead to multiple packages being installed in sum (I'm guessing that's not typically desired), and we generally do not hard code versions in the documentation; they can get stale quite quickly.

I also tried `pkg_add -I autoconf`, which fails, reporting ambiguous selection.

So, what's the best way to approach this, @brad0? If there's a non-intrusive, accepted method to make this more automatic, that would be more consistent with what we have for other systems, but if that is not the OpenBSD preferred way, adding interactive mode `pkg_add` commands to the docs would be fine (`pkg_add autoconf automake ...`).

brad0 commented:
I added a PR https://github.com/HandBrake/HandBrake-docs/pull/170 for the documentation for initial OpenBSD support. I have adjusted everything to OpenBSD's dependencies and ways of doing things.

For OpenBSD we build with our ports tree and build packages.

Currently testing that out it gets most of the build done but trips over build issues with the contrib copies of libbluray and x265.

brad0 commented:
> OpenBSD's `pkg_add` generally works in interactive mode, and with many versions of packages available, I'm not aware of the preferred incantation for installing a suitable set of packages with a one liner. For example, let's install `autoconf`:
> 
> ```
> $ pkg_add autoconf
> #...
> Ambiguous: choose package for autoconf
>         0: <None>
>         1: autoconf-2.13p4
>         2: autoconf-2.52p6
> #...
>         16: autoconf-2.69p3
>         17: autoconf-2.71
> Your choice: 
> ```
> 
> There are seventeen available packages and nothing here indicates `autoconf-2.69p3` is already installed. Presumably, the user should know their system well enough to make a competent decision. Specifying the full package name with version does install without further interruption, but can lead to multiple packages being installed in sum (I'm guessing that's not typically desired), and we generally do not hard code versions in the documentation; they can get stale quite quickly.

Multiple versions of autoconf and automake can be and are supposed to be installed simultaniously. You just have to use the environment variables AUTOCONF_VERSION / AUTOMAKE_VERSION to specific the versions to the autoconf / automake wrappers.
> I also tried `pkg_add -I autoconf`, which fails, reporting ambiguous selection.
> 
> So, what's the best way to approach this, @brad0? If there's a non-intrusive, accepted method to make this more automatic, that would be more consistent with what we have for other systems, but if that is not the OpenBSD preferred way, adding interactive mode `pkg_add` commands to the docs would be fine (`pkg_add autoconf automake ...`).

We usually use our ports tree. But I have provided initial documentation to try and support the standalone build as well. I'm looking into build issues with libbluray and x265. It looks like HandBrake uses a snapshot of x265 past the 3.5 release.

bradleysepos commented:
> I added a PR https://github.com/HandBrake/HandBrake-docs/pull/170 for the documentation for initial OpenBSD support. I have adjusted everything to OpenBSD's dependencies and ways of doing things.

This is excellent, thank you! I will give it a full walkthrough soon.

Re: contrib libraries, you have probably noticed we patch and statically build in some cases where we need to fix issues or add support for emerging features. We try to use system libs where possible, but in some cases it's not (certain hardware encoder support, bugs in unmaintained projects, etc.). Building without our patches usually breaks things in subtle ways, sometimes in not so subtle ways. It's a tough to please everyone.

brad0 commented:
I have filed a PR to fix building on x265.

https://bitbucket.org/multicoreware/x265_git/pull-requests/8/fix-bsd-build

brad0 commented:
Also filed a PR to fix building libbluray out of the box (we had a local and different hack for this).

https://code.videolan.org/videolan/libbluray/-/merge_requests/36

brad0 commented:
All of my libbluray PRs have been commited. I'll see about trying to push for a 1.3.3 release soon-ish.

brad0 commented:
Also to note based on a patch I noticed both in the FreeBSD ports tree and the HandBrake contrib dir for libbluray I pushed a patch to fix a mistake when the FreeBSD JNI header was added.

https://code.videolan.org/videolan/libbluray/-/merge_requests/35

Nomis101 commented:
> I have filed a PR to fix building on x265.
> 
> https://bitbucket.org/multicoreware/x265_git/pull-requests/8/fix-bsd-build

Not sure if it works for x265 with the normal pull request approach. Normally they have this mailing list for patches:
https://bitbucket.org/multicoreware/x265_git/wiki/Contribute

brad0 commented:
> Not sure if it works for x265 with the normal pull request approach. Normally they have this mailing list for patches: https://bitbucket.org/multicoreware/x265_git/wiki/Contribute

Oh, I see. Thanks.

brad0 commented:
> > I have filed a PR to fix building on x265.
> > https://bitbucket.org/multicoreware/x265_git/pull-requests/8/fix-bsd-build
> 
> Not sure if it works for x265 with the normal pull request approach. Normally they have this mailing list for patches: https://bitbucket.org/multicoreware/x265_git/wiki/Contribute

All of that cumbersome CLA for a one line diff.

brad0 commented:
With what is there it now builds.

```
fubar$ ./HandBrakeCLI --version
[03:27:28] Compile-time hardening features are enabled
[03:27:29] hb_init: starting libhb thread
[03:27:29] thread 74aa2b4be40 started ("libhb")
HandBrake 20220925013031-d4dd8f01f-openbsd_support

HandBrake has exited.
```

brad0 commented:
Any chance of having this reviewed?

bradleysepos commented:
Verified building HandBrakeCLI using a basic OpenBSD 7.1 installation. I'm struggling with Gnome, however; it keeps crashing mid-build. Working on it.

brad0 commented:
Ya, I forgot to mention the Gtk+ interface builds. I ran a few encodes from some source content. H264 / HEVC was fine.

bradleysepos commented:
> Ya, I forgot to mention the Gtk+ interface builds. I ran a few encodes from some source content. H264 / HEVC was fine.

I believe you. 😸 Just trying to track down this repeated crash. Gnome is slow and flaky in this VM, regardless. Edit: Seems to be related to parallelism with high core counts.

In any case, this PR should be ready to rebase and merge.

brad0 commented:
The only issue I have noticed so far is encoding with AV1 consistently crashes.

bradleysepos commented:
Reducing jobs from 32 to 8 reduces or avoids the Gnome crashes (still had one). Even so, compilation requires multiple invocations here. I'm not sure it's related to this PR but this one failure seems to indicate an issue spawning processes (observed with `-j8`, even):

```
configure:4330: checking for suffix of executables
configure:4337: /usr/bin/cc -o conftest -g0 -O3 -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2  -Wl,-S -g0 -O3 -fstack-protector-strong conftest.c  >&5
cc: error: unable to execute command: posix_spawn failed: Resource temporarily unavailable
configure:4341: $? = 1
configure:4358: error: in `/home/openbsd/HandBrake/build/gtk':
configure:4360: error: cannot compute suffix of executables: cannot compile and link
```

I initially thought perhaps some weird issue with the mp kernel's scheduler, but I had no issues throwing 32 cores at the build outside of Gnome. I am reminded Xfce is nice. 😸

Anyway, `cd build` and running `env AUTOCONF_VERSION=2.71 AUTOMAKE_VERSION=1.16 CC=cc gmake -j8` a few times seems to work around the problem. `gmake -j8 --keep-going` doesn't seem to allow enough time for things to relax.

bradleysepos commented:
Merged, thanks for your work on this!

brad0 commented:
> Reducing jobs from 32 to 8 reduces or avoids the Gnome crashes (still had one). Even so, compilation requires multiple invocations here. I'm not sure it's related to this PR but this one failure seems to indicate an issue spawning processes (observed with `-j8`, even):
> 
> ```
> configure:4330: checking for suffix of executables
> configure:4337: /usr/bin/cc -o conftest -g0 -O3 -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2  -Wl,-S -g0 -O3 -fstack-protector-strong conftest.c  >&5
> cc: error: unable to execute command: posix_spawn failed: Resource temporarily unavailable
> configure:4341: $? = 1
> configure:4358: error: in `/home/openbsd/HandBrake/build/gtk':
> configure:4360: error: cannot compute suffix of executables: cannot compile and link
> ```
> 
> I initially thought perhaps some weird issue with the mp kernel's scheduler, but I had no issues throwing 32 cores at the build outside of Gnome. I am reminded Xfce is nice. 😸
> 
> Anyway, `cd build` and running `env AUTOCONF_VERSION=2.71 AUTOMAKE_VERSION=1.16 CC=cc gmake -j8` a few times seems to work around the problem. `gmake -j8 --keep-going` doesn't seem to allow enough time for things to relax.



> Reducing jobs from 32 to 8 reduces or avoids the Gnome crashes (still had one). Even so, compilation requires multiple invocations here. I'm not sure it's related to this PR but this one failure seems to indicate an issue spawning processes (observed with `-j8`, even):
> 
> ```
> configure:4330: checking for suffix of executables
> configure:4337: /usr/bin/cc -o conftest -g0 -O3 -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2  -Wl,-S -g0 -O3 -fstack-protector-strong conftest.c  >&5
> cc: error: unable to execute command: posix_spawn failed: Resource temporarily unavailable
> configure:4341: $? = 1
> configure:4358: error: in `/home/openbsd/HandBrake/build/gtk':
> configure:4360: error: cannot compute suffix of executables: cannot compile and link
> ```
> 
> I initially thought perhaps some weird issue with the mp kernel's scheduler, but I had no issues throwing 32 cores at the build outside of Gnome. I am reminded Xfce is nice. 😸
> 
> Anyway, `cd build` and running `env AUTOCONF_VERSION=2.71 AUTOMAKE_VERSION=1.16 CC=cc gmake -j8` a few times seems to work around the problem. `gmake -j8 --keep-going` doesn't seem to allow enough time for things to relax.

With that many jobs you might be hitting default resource limits. I typically add to my .profile..

```
ulimit -d 6291456
ulimit -n 1024
ulimit -p 512
```

and

```
sudo usermod -L staff [login]
```

to set the login to the staff class.

bradleysepos commented:
Thanks for the help. I tried some variations on the above as well as increasing Gnome's memory limit. I did manage one 32-job build successfully but had problems with subsequent builds. 🤷‍♂️ Serial builds work well enough, I'm satisfied.

Nomis101 commented:
> I have filed a PR to fix building on x265.
> 
> https://bitbucket.org/multicoreware/x265_git/pull-requests/8/fix-bsd-build

Finally this has been applied on x265 master.
